rfc/0002/README.md

```diff
@@ -0,0 +1,169 @@
+- Between the AUTHOR and USER: The application expects certain storage, and the #DW must meet those expectations.
+- ADMINISTRATOR to RABBIT: Admin tells Rabbit how to run the containerized application with the required storage.
+- Between USER and RABBIT: User provides the #DW container directive in the job specification. Rabbit validates and interprets the directive.
+
+Proposal
+--------
+
+The proposal below might take a couple of read-throughs; I've also added a concrete example afterward that might help.
+
+1. The AUTHOR writes their application expecting NNF Storage at specific locations. For each storage requirement, they define:
+    1. a unique name for the storage which can be referenced in the 'container' directive
+    2. the expected storage types; if necessary
+    3. the required mount path or mount path prefix
+    4. other constraints or storage requirements (e.g. minimum capacity)
+2. The AUTHOR works with the ADMINISTRATOR to define:
+    1. a unique name for the program to be referred by USER
+    2. the pod specification for executing their program
+    3. the NNF storage requirements described above. 
+3. The ADMINISTRATOR creates a corresponding _NNF Container Profile_ custom kubernetes resource
+4. The USER who desires to use the application works with the AUTHOR and the related Container Profile to understand the storage requirements.
+5. The USER submits a WLM job with the #DW container fields populated
+6. WLM runs the job and drives the job through the following stages...
+    1. Proposal: RABBIT validates the #DW container directive by comparing the supplied values to what is listed in the Container Profile. If the USER fails to meet the requirements, the job fails. 
+    2. Pre-run: RABBIT software will:
+        1. create a config map reflecting the storage requirements and any runtime parameters
+        2. duplicate the pod specification from the Container Profile and patches the necessary Volumes and the config map. The spec is used as the basis for starting the necessary pods and containers.
+    3. The containerized application executes. The expected mounts are available per the requirements and celebration occurs.
+
+Example
+-------
+
+Say I authored a simple application, `foo`, that requires Rabbit local GFS2 storage and a persistent Lustre storage volume. As the author, my program is coded to expect the GFS2 volume is mounted at `/foo/local` and the Lustre volume is mounted at `/foo/persistent`
+
+Working with an administrator, my applications storage requirements and pod specification are placed in an NNF Container Profile `foo`
+
+```yaml
+kind: NnfContainerProfile
+apiVersion: v1alpha1
+metadata:
+    name: foo
+    namespace: default
+spec:
+    storages:
+    - name: $JOB_DW_foo-local-storage
+      type: gfs2
+    - name: $PERSISTENT_DW_foo-persistent-storage
+      type: lustre
+    podSpec:
```

Comment:
Match what the rest of k8s does for templates.  For example see how they do it with a [DaemonSet](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/), where you have /spec/template which is the template for the pod (and thus contains a metadata and spec section of its own).  This will provide consistency with the rest of k8s as well as allow for NnfContainerProfile authors to influence the metadata section of a pod.

Reply:
Can you explain which template fields would be useful for the admin to define? I'd prefer to keep only truly configurable parameters and not expose too much if possible as that only complicates things.

Reply:
There are two meanings of template.  The simple version and hard version.  I was referring to the simple version, which is I think the minimum, but you might be asking about the hard version.

### Simple
User defines a yaml pod document that gets merged with some additive info by Rabbit.  That means a user ought to be able to specify in NnfContainerProfile all things you might see in a [Pod](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/).  The current proposal does not allow for this, since there is no place for user to influence the pod metadata, which is where you might place labels and annotations.

### Sophisticated
In addition to what you have in Simple, letting the user put SYMBOLS in the spec that get replaced by rabbit.  Offhand I'm not sure what a use case would be (though I have seen it, somewhere...).  But the idea is basically, there's a value that won't be known by the user but that rabbit will know at runtime and could fill in.

--

Somewhat related, you should think about how Pods will be named from a NnfContainerProfile.  It should probably match DaemonSet behavior for consistency, since that is what this will be like.  In fact, NnfContainerProfile is mostly a template for making DaemonSets, isn't it?  So the final Pod name probably needs to have the following identifiers in it:

1. Workflow
2. NnfContainerProfile name
3. Ordinal within the daemonset

So I guess rabbit could use the referenced NnfContainerProfile to create a Daemonset with name $workflow-$NnfContainerProfileName which in turn would pump out Pods with names $workflow-$NnfContainerProfileName-$ordinal (or whatever the daemonset controller does).

Reply:
I mean the actual PodTemplateSpec https://github.com/kubernetes/api/blob/master/core/v1/types.go#L4003